### ISSUE TITLE
[0.15] Batch Normalization test failure (clang-6.0)

### ISSUE THREAD

**Issue by cdluminate** [intel/mkl-dnn#284]
```
[omitted hundreds of similar lines]
/home/lumin/packages/mkldnn.pkg/mkl-dnn/tests/gtests/test_batch_normalization.cpp:224: Failure
The difference between (out_diff_src - ref_diff_src) / norm_max and 0. is 0.71124374866485596, which exceeds eps, where
(out_diff_src - ref_diff_src) / norm_max evaluates to 0.71124374866485596,
0. evaluates to 0, and
eps evaluates to 0.62720000743865967.
/home/lumin/packages/mkldnn.pkg/mkl-dnn/tests/gtests/test_batch_normalization.cpp:224: Failure
The difference between (out_diff_src - ref_diff_src) / norm_max and 0. is 0.83767950534820557, which exceeds eps, where
(out_diff_src - ref_diff_src) / norm_max evaluates to 0.83767950534820557,
0. evaluates to 0, and
eps evaluates to 0.62720000743865967.
/home/lumin/packages/mkldnn.pkg/mkl-dnn/tests/gtests/test_batch_normalization.cpp:224: Failure
The difference between (out_diff_src - ref_diff_src) / norm_max and 0. is 0.74958527088165283, which exceeds eps, where
(out_diff_src - ref_diff_src) / norm_max evaluates to 0.74958527088165283,
0. evaluates to 0, and
eps evaluates to 0.62720000743865967.
/home/lumin/packages/mkldnn.pkg/mkl-dnn/tests/gtests/test_batch_normalization.cpp:224: Failure
The difference between (out_diff_src - ref_diff_src) / norm_max and 0. is 0.73271095752716064, which exceeds eps, where
(out_diff_src - ref_diff_src) / norm_max evaluates to 0.73271095752716064,
0. evaluates to 0, and
eps evaluates to 0.62720000743865967.
/home/lumin/packages/mkldnn.pkg/mkl-dnn/tests/gtests/test_batch_normalization.cpp:224: Failure
The difference between (out_diff_src - ref_diff_src) / norm_max and 0. is 0.7993740439414978, which exceeds eps, where
(out_diff_src - ref_diff_src) / norm_max evaluates to 0.7993740439414978,
0. evaluates to 0, and
eps evaluates to 0.62720000743865967.
/home/lumin/packages/mkldnn.pkg/mkl-dnn/tests/gtests/test_batch_normalization.cpp:224: Failure
The difference between (out_diff_src - ref_diff_src) / norm_max and 0. is 0.71368342638015747, which exceeds eps, where
(out_diff_src - ref_diff_src) / norm_max evaluates to 0.71368342638015747,
0. evaluates to 0, and
eps evaluates to 0.62720000743865967.
/home/lumin/packages/mkldnn.pkg/mkl-dnn/tests/gtests/test_batch_normalization.cpp:224: Failure
The difference between (out_diff_src - ref_diff_src) / norm_max and 0. is 0.7133939266204834, which exceeds eps, where
(out_diff_src - ref_diff_src) / norm_max evaluates to 0.7133939266204834,
0. evaluates to 0, and
eps evaluates to 0.62720000743865967.
/home/lumin/packages/mkldnn.pkg/mkl-dnn/tests/gtests/test_batch_normalization.cpp:224: Failure
The difference between (out_diff_src - ref_diff_src) / norm_max and 0. is 0.80129843950271606, which exceeds eps, where
(out_diff_src - ref_diff_src) / norm_max evaluates to 0.80129843950271606,
0. evaluates to 0, and
eps evaluates to 0.62720000743865967.
/home/lumin/packages/mkldnn.pkg/mkl-dnn/tests/gtests/test_batch_normalization.cpp:224: Failure
The difference between (out_diff_src - ref_diff_src) / norm_max and 0. is 0.73182350397109985, which exceeds eps, where
(out_diff_src - ref_diff_src) / norm_max evaluates to 0.73182350397109985,
0. evaluates to 0, and
eps evaluates to 0.62720000743865967.
/home/lumin/packages/mkldnn.pkg/mkl-dnn/tests/gtests/test_batch_normalization.cpp:224: Failure
The difference between (out_diff_src - ref_diff_src) / norm_max and 0. is 0.75081086158752441, which exceeds eps, where
(out_diff_src - ref_diff_src) / norm_max evaluates to 0.75081086158752441,
0. evaluates to 0, and
eps evaluates to 0.62720000743865967.
/home/lumin/packages/mkldnn.pkg/mkl-dnn/tests/gtests/test_batch_normalization.cpp:224: Failure
The difference between (out_diff_src - ref_diff_src) / norm_max and 0. is 0.77015256881713867, which exceeds eps, where
(out_diff_src - ref_diff_src) / norm_max evaluates to 0.77015256881713867,
0. evaluates to 0, and
eps evaluates to 0.62720000743865967.
/home/lumin/packages/mkldnn.pkg/mkl-dnn/tests/gtests/test_batch_normalization.cpp:224: Failure
The difference between (out_diff_src - ref_diff_src) / norm_max and 0. is 0.7214093804359436, which exceeds eps, where
(out_diff_src - ref_diff_src) / norm_max evaluates to 0.7214093804359436,
0. evaluates to 0, and
eps evaluates to 0.62720000743865967.
[  FAILED  ] GoogleNet_NCHW/bnrm_test_float.TestsBnrm/1, where GetParam() = 40-byte object <01-00 00-00 05-00 00-00 05-00 00-00 02-00 00-00 40-00 00-00 01-00 00-00 38-00 00-00 38-00 00-00 AC-C5 27-37 04-00 00-00> (5665 ms)
```

-----------------------------------------------------------------------------

### Environment
Intel MKL-DNN includes hardware-specific optimizations and may behave
differently on depending on the compiler and build environment. Include
the following information to help reproduce the issue:
* CPU make and model (try `lscpu`; if your `lscpu` does not list CPU flags,
  try running `cat /proc/cpuinfo | grep flags | sort -u`)

```
Architecture:        x86_64
CPU op-mode(s):      32-bit, 64-bit
Byte Order:          Little Endian
CPU(s):              4
On-line CPU(s) list: 0-3
Thread(s) per core:  1
Core(s) per socket:  4
Socket(s):           1
NUMA node(s):        1
Vendor ID:           GenuineIntel
CPU family:          6
Model:               158
Model name:          Intel(R) Core(TM) i5-7440HQ CPU @ 2.80GHz
Stepping:            9
CPU MHz:             2803.253
CPU max MHz:         3800.0000
CPU min MHz:         800.0000
BogoMIPS:            5616.00
Virtualization:      VT-x
L1d cache:           32K
L1i cache:           32K
L2 cache:            256K
L3 cache:            6144K
NUMA node0 CPU(s):   0-3
Flags:               fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf tsc_known_freq pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb invpcid_single pti ibrs ibpb stibp tpr_shadow vnmi flexpriority ept vpid fsgsbase tsc_adjust bmi1 hle avx2 smep bmi2 erms invpcid rtm mpx rdseed adx smap clflushopt intel_pt xsaveopt xsavec xgetbv1 xsaves dtherm ida arat pln pts hwp hwp_notify hwp_act_window hwp_epp
```

* OS version (`uname -a`)

```
Linux Asuna 4.17.0-1-amd64 #1 SMP Debian 4.17.6-2 (2018-07-15) x86_64 GNU/Linux
```

* Compiler version (`gcc --version`)

The compiler is Clang shipped by Debian unstable

```
~ ❯❯❯ clang --version
clang version 6.0.1-2 (tags/RELEASE_601/final)
Target: x86_64-pc-linux-gnu
Thread model: posix
InstalledDir: /usr/bin
~ ❯❯❯ clang++ --version
clang version 6.0.1-2 (tags/RELEASE_601/final)
Target: x86_64-pc-linux-gnu
Thread model: posix
InstalledDir: /usr/bin
```

* MKLROOT value (`echo MKLROOT=$MKLROOT`)

`none`

* CMake version (`cmake --version`)

`cmake version 3.11.2`

* CMake output log

```
dh_auto_configure -- -DCMAKE_BUILD_TYPE=Release \
	-DCMAKE_CXX_COMPILER=clang++ \
	-DCMAKE_C_COMPILER=clang
	cd obj-x86_64-linux-gnu && cmake -DCMAKE_INSTALL_PREFIX=/usr -DCMAKE_VERBOSE_MAKEFILE=ON -DCMAKE_BUILD_TYPE=None -DCMAKE_INSTALL_SYSCONFDIR=/etc -DCMAKE_INSTALL_LOCALSTATEDIR=/var -DCMAKE_EXPORT_NO_PACKAGE_REGISTRY=ON -DCMAKE_FIND_PACKAGE_NO_PACKAGE_REGISTRY=ON -DCMAKE_INSTALL_RUNSTATEDIR=/run "-GUnix Makefiles" -DCMAKE_BUILD_TYPE=Release -DCMAKE_CXX_COMPILER=clang\+\+ -DCMAKE_C_COMPILER=clang ..
CMake Deprecation Warning at CMakeLists.txt:21 (cmake_policy):
  The OLD behavior for policy CMP0048 will be removed from a future version
  of CMake.

  The cmake-policies(7) manual explains that the OLD behaviors of all
  policies are deprecated and that a policy should be set to OLD only under
  specific short-term circumstances.  Projects should be ported to the NEW
  behavior and not rely on setting a policy to OLD.


CMake Deprecation Warning at CMakeLists.txt:22 (cmake_policy):
  The OLD behavior for policy CMP0054 will be removed from a future version
  of CMake.

  The cmake-policies(7) manual explains that the OLD behaviors of all
  policies are deprecated and that a policy should be set to OLD only under
  specific short-term circumstances.  Projects should be ported to the NEW
  behavior and not rely on setting a policy to OLD.


-- The C compiler identification is Clang 6.0.1
-- The CXX compiler identification is Clang 6.0.1
-- Check for working C compiler: /usr/bin/clang
-- Check for working C compiler: /usr/bin/clang -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Check for working CXX compiler: /usr/bin/clang++
-- Check for working CXX compiler: /usr/bin/clang++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Detecting Intel(R) MKL: trying mklml_intel
-- Detecting Intel(R) MKL: trying mklml
-- Detecting Intel(R) MKL: trying mkl_rt
CMake Warning at cmake/MKL.cmake:196 (message):
  Intel(R) MKL not found.  Some performance features may not be available.
  Please run scripts/prepare_mkl.sh to download a minimal set of libraries or
  get a full version from https://software.intel.com/en-us/intel-mkl
Call Stack (most recent call first):
  cmake/OpenMP.cmake:25 (include)
  CMakeLists.txt:61 (include)


-- Found OpenMP_C: -fopenmp=libomp (found version "3.1") 
-- Found OpenMP_CXX: -fopenmp=libomp (found version "3.1") 
-- Found OpenMP: TRUE (found version "3.1")  
-- Found Doxygen: /usr/bin/doxygen (found version "1.8.13") found components:  doxygen dot 
-- VTune profiling environment is unset
-- Looking for pthread.h
-- Looking for pthread.h - found
-- Looking for pthread_create
-- Looking for pthread_create - found
-- Found Threads: TRUE  
-- Configuring done
-- Generating done
CMake Warning:
  Manually-specified variables were not used by the project:

    CMAKE_EXPORT_NO_PACKAGE_REGISTRY
    CMAKE_INSTALL_LOCALSTATEDIR
    CMAKE_INSTALL_RUNSTATEDIR
    CMAKE_INSTALL_SYSCONFDIR


-- Build files have been written to: /home/lumin/packages/mkldnn.pkg/mkl-dnn/obj-x86_64-linux-gnu
make[1]: Leaving directory '/home/lumin/packages/mkldnn.pkg/mkl-dnn'
   dh_auto_build -O--buildsystem=cmake
	cd obj-x86_64-linux-gnu && make -j4 "INSTALL=install --strip-program=true"
```

* git hash (`git log -1 --format=%H`)

`0.15`

**Comment by cdluminate** [comment]
Similar to https://github.com/intel/mkl-dnn/issues/280

**Comment by emfomenk** [comment]
Hi @cdluminate,

I cannot reproduce the failure on my side (seems like a race condition or something that doesn't always reproduce the issue). 

Could you please run the test (maybe few times) with `OMP_NUM_THREADS=1` environment variable, to check that the problem is in threading?

**Comment by cdluminate** [comment]
@emfomenk After applying that patch I cannot reproduce the failure. I'll report here if I encountered this problem in the future builds.

**Comment by emfomenk** [comment]
Thanks for checking @cdluminate!
Will try to promote the fix as soon as possible.

**Comment by vpirogov** [comment]
Fixed by d352e0db707cb943b151407353d0d6ed0286a226

**Comment by cdluminate** [comment]
Confirm that I cannot repro the bug anymore with that fix.
